pkg/pool-weighted/test/ManagedPool.test.ts

```diff
@@ -95,30 +103,158 @@ describe('ManagedPool', function () {
 
   describe('swap', () => {
     sharedBeforeEach('deploy pool', async () => {
-      pool = await deployPool({ vault: undefined, swapEnabledOnStart: true });
+      pool = await deployPool({ swapEnabledOnStart: true });
 
       await pool.init({ from: other, initialBalances });
     });
 
-    context('when swaps are disabled', () => {
-      sharedBeforeEach('deploy pool', async () => {
-        await pool.setSwapEnabled(owner, false);
+    context('token swaps', () => {
+      context('when swaps are disabled', () => {
+        sharedBeforeEach('deploy pool', async () => {
+          await pool.setSwapEnabled(owner, false);
+        });
+
+        it('it reverts', async () => {
+          await expect(
+            pool.swapGivenIn({ in: 1, out: 2, amount: fp(0.1), from: other, recipient: other })
+          ).to.be.revertedWith('SWAPS_DISABLED');
+        });
       });
 
-      it('it reverts', async () => {
-        await expect(pool.swapGivenIn({ in: 1, out: 0, amount: fp(0.1) })).to.be.revertedWith('SWAPS_DISABLED');
+      context('when swaps are enabled', () => {
+        sharedBeforeEach('deploy pool', async () => {
+          await pool.setSwapEnabled(owner, true);
+        });
+
+        it('swaps are not blocked', async () => {
+          await expect(pool.swapGivenIn({ in: 1, out: 2, amount: fp(0.1), from: other, recipient: other })).to.not.be
+            .reverted;
+        });
       });
     });
 
-    context('when swaps are enabled', () => {
-      sharedBeforeEach('deploy pool', async () => {
-        await pool.setSwapEnabled(owner, true);
+    context('join swaps', () => {
+      function itPerformsAJoinSwapCorrectly(
+        doJoinSwap: () => Promise<SwapResult>,
+        queryJoinSwap: () => Promise<BigNumber[]>,
+        queryEquivalentJoin: () => Promise<JoinQueryResult>
+      ) {
+        function isEquivalentToARegularJoin() {
+          it("doesn't revert", async () => {
+            await expect(doJoinSwap()).to.not.be.reverted;
+          });
+
+          it('returns the same amount of BPT as the equivalent join', async () => {
+            const joinSwapResult = await queryJoinSwap();
+            const joinResult = await queryEquivalentJoin();
+
+            // BPT is leaving the Vault and so is represented as a negative value.
+            expect(joinSwapResult[0].mul(-1)).to.be.eq(joinResult.bptOut);
+          });
+
+          it('takes the same amount of tokens as the equivalent join', async () => {
+            const joinSwapResult = await queryJoinSwap();
+            const joinResult = await queryEquivalentJoin();
+
+            expect(joinSwapResult[1]).to.be.eq(joinResult.amountsIn[2]);
```

Comment:
Can we use contants for some of these magical indices? Specially token indices that are shifted due to :musical_score:

Reply:
Or even better, receive them as arguments

Reply:
It's a bit complex as we have 3 different types of indexes floating about but hopefully it should be fairly clear now.